lib/Enforcers/RevngEnforcers.cpp

```diff
@@ -0,0 +1,102 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/StringMap.h"
+
+#include "revng/AutoEnforcer/AutoEnforcerLibraryRegistry.h"
+#include "revng/AutoEnforcer/AutoEnforcerTarget.h"
+#include "revng/AutoEnforcer/InspectableLLVMContainer.h"
+#include "revng/Enforcers/CompileModule.h"
+#include "revng/Enforcers/Isolate.h"
+#include "revng/Enforcers/Lift.h"
+#include "revng/Enforcers/LinkForTranslation.h"
+#include "revng/Enforcers/LinkSupport.h"
+#include "revng/Enforcers/RevngEnforcers.h"
+#include "revng/Enforcers/RootKind.h"
+#include "revng/Enforcers/TemporaryFile.h"
+
+using namespace std;
+using namespace AutoEnforcer;
+
+Granularity AutoEnforcer::RootGranularity("Root Granularity");
+Granularity
+  AutoEnforcer::FunctionsGranularity("Function Granularity", RootGranularity);
+
+Kind AutoEnforcer::Binary("Binary", &RootGranularity);
+
+Kind AutoEnforcer::Object("Object", &RootGranularity);
+Kind AutoEnforcer::Translated("Translated", &RootGranularity);
+Kind AutoEnforcer::Dead("Dead", &RootGranularity);
+
+Kind AutoEnforcer::CFepper("CFepper", &FunctionsGranularity);
+Kind AutoEnforcer::ABIEnforced("ABIEnforced", Isolated, &FunctionsGranularity);
+
+llvm::Error RevngModelAutoEnforcerWrapper::save(llvm::StringRef Path) const {
+  return serializeToFile(Model.getReadOnlyModel(), Path);
+}
+
+llvm::Error RevngModelAutoEnforcerWrapper::load(llvm::StringRef Path) {
+  if (not llvm::sys::fs::exists(Path))
+    return llvm::Error::success();
+
+  auto MaybeModel = TupleTree<model::Binary>::loadFromFile(Path);
+  if (not MaybeModel)
+    return llvm::make_error<llvm::StringError>("Could not parse model",
+                                               MaybeModel.getError());
+
+  Model = std::move(*MaybeModel);
+  return llvm::Error::success();
+}
+namespace AutoEnforcer {
+llvm::Error autoEnforcerRevngConfigurationCallback(const PipelineLoader &Loader,
+                                                   LLVMEnforcer &NewPass) {
+  using Wrapper = RevngModelAutoEnforcerWrapper;
+  auto MaybeModelWrapper = Loader.getContext().get<Wrapper>(Wrapper::Name);
+  if (not MaybeModelWrapper)
+    return MaybeModelWrapper.takeError();
+
+  auto &ModelWrapper = (*MaybeModelWrapper)->getModelWrapper();
+  NewPass.emplacePass<LoadModelEnforcerPass>(ModelWrapper);
+  return llvm::Error::success();
+}
+
+const char RevngModelAutoEnforcerWrapper::ID = '0';
+
+class RevngRegistryLibrary : public AutoEnforcerLibraryRegistry {
+
+public:
+  void registerContainersAndEnforcers(PipelineLoader &Loader) override {
+    Loader.registerContainerFactory<TemporaryFileFactory>("Binary", Binary);
+    Loader.registerContainerFactory<TemporaryFileFactory>("Object", Object);
+    Loader.registerContainerFactory<TemporaryFileFactory>("Translated",
+                                                          Translated);
+    Loader.registerEnforcer("Lift", LiftEnforcer());
```

Comment:
Do enforcers have to be default-constructible?
If so,
```suggestion
    Loader.registerEnforcer<LiftEnforcer>("Lift");
```
Also, are you sure we don't want the name to be a property of the enforcer?
Why put it here?

Reply:
we have both currently. i used the other overload there which is indeed nicer there.